### PR TITLE
Master easy filters

### PR DESCRIPTION
**Pull request by ged-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by celm1990** [comment]
Hi, great feature, but please take into account following suggestion.
When **add custom group**, options(day, week month, year,  etc) not working

image

image

**Comment by ged-odoo** [comment]
Hello @celm1990 
Thanks for commenting. Do not worry, this is still a work in progress. We will try to fix every bugs

File: addons/web/static/src/js/views/search/search_view.js [hunk 1]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
         fv.arch = utils.xml_to_json(doc, true);
         return fv;
     },
+    /**
+     * @private
+     * @param {string[]]} selectedGroupIds
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: you should at least mention what are these groupids...

File: addons/web/static/src/js/views/search/search_view.js [hunk 2]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+                }
+                return acc;
+            },
+            []
+        );
+        this.selectedGroupIds.groupByCategory = selectedGroupIds;
+        this.groupby_menu.unsetGroups(groupIds);
+    },
+    /**
+     * @private
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: missing explanation

File: addons/web/static/src/js/views/search/search_view.js [hunk 3]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+        this.selectedGroupIds.filterCategory = selectedGroupIds;
+        this.filters_menu.unsetGroups(groupIds);
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     *
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: missing explanation

File: addons/web/static/src/js/views/search/search_view.js [hunk 4]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+            group.toggle(filter);
+        }
+    },
+    /**
+     * @private
+     * @param {OdooEvent} event
+     */
+    _onNewGroupby: function (event) {
+        var attrs = {
+                context:"{'group_by':'" + event.data.fieldName + "''}",
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: weird indenting

File: addons/web/static/src/js/views/search/search_view.js [hunk 5]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+                name: event.data.description,
+            };
+
+        var groupby = new search_inputs.Filter({attrs: attrs}, this);
+        if (event.data.optionId) {
+            // _.extend(attrs,  {interval: event.data.optionId.split('__')[2]});
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: remove this

File: addons/web/static/src/js/views/search/search_view.js [hunk 6]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+            group.toggle(groupby);
+        }
+        if (event.data.category === 'filterCategory') {
+            var filter = _.findWhere(this.filtersMapping, {filterId: event.data.itemId}).filter;
+            group = _.findWhere(this.groupsMapping, {groupId: event.data.groupId}).group;
+            group.toggle(filter);
+        }
+    },
+    /**
+     * @private
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: missing explanation

File: addons/web/static/src/js/views/search/search_view.js [hunk 7]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+            groupby: groupby,
+            groupId: event.data.groupId,
+        });
+        this.groupsMapping.push({
+            groupId: event.data.groupId,
+            group: group,
+            category: 'Group By',
+        });
+    },
+        /**
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: indentation

File: addons/web/static/src/js/views/search/search_view.js [hunk 8]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+     * @param {OdooEvent} event
+     */
+    _onNewFilter: function (event) {
+        var self= this;
+        var filter;
+        var filters = [];
+        var groupId;
+
+        _.each(event.data, function (filterDescription) {
+                filter = new search_inputs.Filter(filterDescription.filter, this);
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: indentation

File: addons/web/static/src/js/views/search/search_view.js [hunk 9]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+                if (!groupId) {
+                    groupId = filterDescription.groupId;
+                }
+            });
+        var group = new search_inputs.FilterGroup(filters, this, this.intervalMapping);
+        filters.forEach(function (filter) {
+            group.toggle(filter, {silent: true});
+        });
+        this.query.trigger('reset');
+
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: empty line? no need for 2 lines

File: addons/web/static/src/js/views/search/search_view.js [hunk 10]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+
+        this.groupsMapping.push({
+            groupId: groupId,
+            group: group,
+            category: 'Filters',
+        });
+    },
+    /**
+     *
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: missing explanation

File: addons/web/static/src/js/views/search/search_view.js [hunk 11]
```diff
@@ -699,6 +853,154 @@ var SearchView = Widget.extend({
...
+                return acc;
+            },
+            []
+        );
+        this.selectedGroupIds.groupByCategory = selectedGroupIds;
+        this.groupby_menu.unsetGroups(groupIds);
+    },
+    /**
+     * @private
+     * @param {string[]]} selectedGroupIds
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/search/search_view.js]: what are these?

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    events: {
+        'click li.o_menu_item': '_onItemClick',
+        'click li.o_item_option': '_onOptionClick',
+        'click li.o_menu_item span.o_submenu_switcher': '_openItemOptions',
+        'click span.o_trash_button': '_onTrashClick',
+        'hidden.bs.dropdown': '_onBootstrapClose',
+    },
+
+    /**
+     * @param {Widget} parent
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: missing @ override

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+     * @param {Object} dropdowHeader object with keys 'title' and 'icon'
+     * @param {Object} items values are arrays of items (see interface IItems)
+     * @param {Object} itemGenerator
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: explain what this is

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    init: function (parent, dropdownHeader, items) {
+        this._super(parent);
+        this.dropdownCategory = dropdownHeader.category;
+        this.dropdownTitle = dropdownHeader.title;
+        this.dropdownIcon = dropdownHeader.icon;
+        this.dropdownSymbol = dropdownHeader.symbol || 'caret';
+        this.items = _.sortBy(items, 'groupId');
+        _.each(this.items, this._prepareItem.bind(this));
+    },
+
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: empty line,m issing docstring

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * @param {string} cssclass
+     */
+    setElStyle: function (cssclass) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: remove this

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        this.$el.removeClass('btn-group')
+            .addClass(cssclass);
+     },
+    /**
+     * @param {string} cssclass
+     */
+    setMainButtonStyle: function (cssclass) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: remove this

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+        this.$el.find('button:first').removeClass('o_dropdown_toggler_btn btn btn-sm btn-default dropdown-toggle')
+            .addClass(cssclass);
+     },
+
+    /**
+     * @param {Array[number]} groupIds
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: number? not string? also, missind documentation

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+        this.items.splice(itemIndex, 1);
+        this._renderMenuItems();
+        var eventData = {
+            category: this.dropdownCategory,
+            itemId: itemId,
+            groupId: item.groupId,
+        };
+        this.trigger_up('menu_item_deleted', eventData);
+    },
+
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: empty line

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+    /**
+     * @private
+     */
+    _renderMenuItems: function () {
+        var newMenuItems = QWeb.render('DropdownMenu.MenuItems', {widget: this});
+        this.$el.find('.o_menu_item,.divider[data-removable="1"]').remove();
+        this.$('ul.o_dropdown_menu').prepend($(newMenuItems));
+    },
+    /**
+     * @private
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: maybe add more explanation

File: addons/web/static/src/js/widgets/dropdown_menu.js (hunk 17 above)
Comment (ged-odoo) [reply addons/web/static/src/js/widgets/dropdown_menu.js]: also, this should be typed number[] (or rather string[]

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+    _toggleMenuOptions: function (itemId) {
+        var item = _.findWhere(this.items, {itemId: itemId});
+        item.isOpen = !item.isOpen;
+        this._renderMenuItems();
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+    /**
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: add empty line

File: addons/web/static/src/js/widgets/dropdown_menu.js [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+        event.preventDefault();
+        event.stopPropagation();
+        var itemId = $(event.currentTarget).data('id');
+        this._toggleMenuOptions(itemId);
+    },
+});
+
+return DropdownMenu;
+
+});
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/dropdown_menu.js]: enter

File: addons/web/static/tests/widgets/dropdown_menu_tests.js [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+
+function createDropdownMenu(dropdownTitle, groups, params) {
+    params = params || {};
+    var target = params.debug ? document.body :  $('#qunit-fixture');
+    var menu = new DropdownMenu(null, dropdownTitle, groups);
+    testUtils.addMockEnvironment(menu, params);
+    menu.appendTo(target);
+    return menu;
+}
+
```
Comment (ged-odoo) [inline addons/web/static/tests/widgets/dropdown_menu_tests.js]: empty line

File: addons/web/static/tests/widgets/dropdown_menu_tests.js [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+
+QUnit.module('Web', {
+    beforeEach: function () {
+        this.items = [
+                {
```
Comment (ged-odoo) [inline addons/web/static/tests/widgets/dropdown_menu_tests.js]: indentation

File: addons/web/static/src/js/views/search/search_inputs.js [hunk 24]
```diff
@@ -515,7 +515,7 @@ var FilterGroup = Input.extend(/** @lends instance.web.search.FilterGroup# */{
      * @param {Array<instance.web.search.Filter>} filters elements of the group
      * @param {instance.web.SearchView} parent parent in which the filters are contained
      */
-    init: function (filters, parent) {
+    init: function (filters, parent, intervalMapping) {
```
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/src/js/views/search/search_inputs.js]: Start here

**Comment by adr-odoo** [comment]
Merged: 3ebff1f4721dd4a9057db4d26162c57c8506700a

**Pull request closed without merging**